### DISCUSSION TITLE
How to finish docker-compose action and mark as succes

### DISCUSSION THREAD

**Discussion by valentijnscholten** [community/community#26852] [Actions] [answered]
<p>Hi,</p>
<p>I have a pyhon project that runs a couple of containers: nginx, uwsgi, mysql, celery, rabbitmq.<br>
And another container to run integration tests (with chromedriver).<br>
This all works fine as docker-compose is supported by GitHub Actions via:</p>
<pre><code class="lang-auto">name: Docker Compose Actions Workflow
on: push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set integration-test mode
        run: docker/setEnv.sh integration_tests
      - name: Build the stack
        run: docker-compose up --build
</code></pre>
<p>This builds the containers, starts them and runs the integration tests as part of the entrypoint script of the integration-tests container.</p>
<p>But after finishing the tests, all containers keep running, or at leasts the ones running a service such as nginx, uwsgi, celery, mysql.<br>
How can I mark the GitHub Action (step) as successful after completing the integration tests. And make sure the containers are stopped and the action continues?</p>
<p>Valentijn</p>

**Comment by airtower-luna** [comment]
<p><code>docker-compose up</code> will by default stay attached to the started container, so it won’t terminate as long as they are running. In principle the way to get around it is to use <code>docker-compose up -d</code> to detach. The tricky thing is that with <code>-d</code> the <code>up</code> command will terminate very quickly (as soon as the containers are up), likely before your tests are complete.</p>
<p>So you could use a sequence like this:</p>
<pre><code class="lang-yaml">      - name: Build the stack
        run: docker-compose up --build -d
      - name: Get test result
        run: |
          # wait for the test results, and terminate with 
          # a non-zero exit code if they are bad
      - name: Stop containers
        # the containers should be stopped regardless of 
        # the test result
        if: always()
        run: docker-compose down
</code></pre>

**Comment by valentijnscholten** [comment]
<p>My first try was with ‘up -d’ which kills the containers too soon indeed. Didn’t want to with some strange dev/null tail in my Dockerfile as suggested elsewhere.<br>
I now have added</p>
<pre><code class="lang-auto"> - name: Logs
       run: docker-compose logs -f integrationtest
</code></pre>
<p>Which seems to correctly wait until the integrationtest container finishes. That container exist/fails with exit code 1 due to a failing test. But the github action is marked as succesful. Not sure how this works, the whole docker-compose setup seems to be undocumented for GitHub actions? <a href="https://github.community/t/github-actions-as-status-checks/16666" class="inline-onebox">Github Actions as Status checks?</a></p>
<p>Now trying your suggestion, build takes 30 minutes. Not sure what the single ‘|’ command does?</p>

**Comment by valentijnscholten** [comment]
<p>Tried with the suggested '|'command, but that doesn’t seem to do anything but instantly complete. So containers are killed, no integration tests are run. Probably need to use docker inspect or something to get the exit code.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="valentijnscholten" data-post="3" data-topic="134224">
<div class="title">
<div class="quote-controls"></div>
 valentijnscholten:</div>
<blockquote>
<p>Which seems to correctly wait until the integrationtest container finishes. That container exist/fails with exit code 1 due to a failing test. But the github action is marked as succesful.</p>
</blockquote>
</aside>
<p>That’s probably because reading the log (what <code>docker-compose logs</code> does) was successful. That’s why I had the “wait for the test results, and terminate with a non-zero exit code if they are bad” comment in my example. I assumed you had some process to retrieve the result from the container (e.g. parse a test log or something).</p>
<p>If you just need the exit code: I’ve done something similar before where I was running the test container independent of <code>docker-compose</code>, and just attaching it to the same virtual network so it could talk to the service containers. That way I could just use the exit code from the <code>docker run</code> command. Depending on how tightly your tests are coupled with the network structure you may be able to do the same by publishing ports and accessing those.</p>
<aside class="quote no-group" data-username="valentijnscholten" data-post="3" data-topic="134224">
<div class="title">
<div class="quote-controls"></div>
 valentijnscholten:</div>
<blockquote>
<p>Not sure what the single ‘|’ command does?</p>
</blockquote>
</aside>
<p>That’s not a command, that’s YAML syntax to start a text block that preserves line breaks. That way you can have more than a one line command in a single <code>run</code> step. :wink:</p>
<p>Example from one of my workflows:</p>
<pre><code class="lang-yaml">      - name: network overview
        run: |
          ip addr show
          cat /etc/hosts
</code></pre>
<p>For more on YAML multiline text, take a look at:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://yaml-multiline.info/favicon.ico" class="site-icon" width="16" height="16">
      <a href="https://yaml-multiline.info/" target="_blank" rel="noopener">yaml-multiline.info</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="20" height="20">

<h3><a href="https://yaml-multiline.info/" target="_blank" rel="noopener">YAML Multiline Strings</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by valentijnscholten** [comment]
<p>Actually I am migrating from Travis and there we are running the integration tests not in a container, but in the main travis script and connecting to the exposes port 8080. I could do the same here I guess and just let the exit code cascade from there. Hopefully it works :slight_smile:</p>

**Comment by airtower-luna** [comment]
<p>That sounds promising, I’d like to hear if it works! :slightly_smiling_face:</p>

**Comment by valentijnscholten** [comment, marked as answer]
<p>I decided to keep the tests running in the container. But started all other containers first with <code>up -d</code> and then start the container with the tests with ‘up --exit-code-from’</p>
<pre><code class="lang-auto">      - name: Integration tests
        run: docker-compose up --exit-code-from integrationtest integrationtest
</code></pre>
<p>this works and the exit code is cascaded nicely into the GitHub action.</p>

**Reply by PRASADREDDY253** [reply]
Its also worked fine for me .Thanks
